src/server/project.ts

```diff
@@ -2215,6 +2215,20 @@ export abstract class Project implements LanguageServiceHost, ModuleResolutionHo
         return this.noDtsResolutionProject;
     }
 
+    /** @internal */
+    updateTargetFile(rootFile: string, targetFileText: string, pastedText: string): { updatedFile: SourceFile | undefined; updatedProgram: Program | undefined; originalProgram: ts.Program | undefined; } {
+        const originalProgram = this.program;
+        this.getScriptInfo(rootFile)?.editContent(0, targetFileText.length, pastedText);
+        this.updateGraph();
+        return { updatedFile: (this.program?.getSourceFile(rootFile)), updatedProgram: this.program, originalProgram };
+    }
+
+    /** @internal */
+    revertUpdatedFile(rootFile: string, updatedText: string, originalText: string) {
+        this.getScriptInfo(rootFile)?.editContent(0, updatedText.length, originalText);
+        this.updateGraph();
+    }

```

Comment:
Two observations:

1. These two methods are the same, except one returns nothing. Just `updateTargetFile` would suffice.
2. You don’t need to pass in the current text of the file; you can get that from the SourceFile or ScriptInfo. You only need to pass in the desired text.

Combining this with my previous comment about needing a try/finally, maybe it would be best to make that part of the API here, so it’s impossible to forget to revert or handle an error. You can make an API like

```ts
host.runWithTemporaryFileUpdate(sourceFile, newText, (updatedProgram, updatedFile) => {
  doEverything();
});
```

I think that would be much better than having two separate methods that must be called symmetrically to avoid catastrophic results 😅